DataDog/datadog-operator
# Support runtimeClassName override (including in DAP)

### What does this PR do?

* Implements runtimeClassName override for pods: 88ff9264195c7ece4e7de742922e533e2865ee4d
* Implements support for runtimeClassName in DatadogAgentProfile: a93199bcb14532cd043d42996b7391a09d32caf1

### Motivation

* Internal team request to support a subset of Agent pods using a different runtimeclass

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: vX.Y.Z
* Cluster Agent: vX.Y.Z

### Describe your test plan
---
**First commit**: override
---

1. Create a runtimeclass (`test-handler` handler exists in `kind` by default) on a `kind` cluster
```yaml
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  # The name the RuntimeClass will be referenced by.
  # RuntimeClass is a non-namespaced resource.
  name: foo
# The name of the corresponding CRI configuration
handler: test-handler
```
2. Provide the override to all components:
```
spec.override.nodeAgent.runtimeClassName: foo
spec.override.clusterAgent.runtimeClassName: foo
```
3. Verify pods are scheduled with said runtimeclass
---
**Second commit**: DAP override
---
1. Create a runtimeclass like for first commit on a 2+ nodes `kind` cluster
2. Label a node with `foo=bar`
3. Deploy generic DDA (no override for runtimeclassname)
4. Deploy the profile:
```yaml
apiVersion: datadoghq.com/v1alpha1
kind: DatadogAgentProfile
metadata:
  name: dap-runtime-foo
spec:
  profileAffinity:
    profileNodeAffinity:
      - key: foo
        operator: In
        values:
          - bar
  config:
    override:
      nodeAgent:
        runtimeClassName: foo
```
5. Verify that a pod on your labelled node is using said specific class instead of default
```shell
╰─❯ kubectl get pods -o custom-columns="NAME:.metadata.name,NODE:.spec.nodeName,RUNTIMECLASS:.spec.runtimeClassName"
NAME                                                      NODE                      RUNTIMECLASS
datadog-agent-6dgr2                                       local-k8s-control-plane   <none>
datadog-agent-vvwd7                                       local-k8s-worker          <none>
datadog-agent-with-profile-system-dap-runtime-foo-flnhm   local-k8s-worker2         foo
datadog-cluster-agent-69b86c4cb9-mjkkn                    local-k8s-worker2         <none>
datadog-operator-manager-696754f84b-9bkf4                 local-k8s-worker          <none>
```
6. Remove the profile and ensure a new Agent pod with default runtimeclass is scheduled


### Checklist
